README.md
# FxHey!

[![Build status](https://gitlab.com/philbooth/FxHey/badges/master/pipeline.svg)](https://gitlab.com/philbooth/FxHey/pipelines)
[![Package status](https://img.shields.io/npm/v/fxhey.svg)](https://www.npmjs.com/package/fxhey)
[![Downloads](https://img.shields.io/npm/dm/fxhey.svg)](https://www.npmjs.com/package/fxhey)
[![License](https://img.shields.io/npm/l/fxhey.svg)](https://opensource.org/licenses/MIT)

Live train announcements for [Firefox Accounts](https://accounts.firefox.com/).

* [You what?](#you-what)
* [How do I install it?](#how-do-i-install-it)
* [How do I use it?](#how-do-i-use-it)
  * [How do I load the library?](#how-do-i-load-the-library)
  * [How do I register for announcements?](#how-do-i-register-for-announcements)
  * [How do I cancel future announcements?](how-do-i-cancel-future-announcements)
  * [What options can I specify?](#what-options-can-i-specify)
* [How do I set up the dev environment?](#how-do-i-set-up-the-dev-environment)
* [What license is it released under?](#what-license-is-it-released-under)

## You what?

Even though I'm
a fully paid-up member
of the dev team,
a lot of the time
I forget where we are
in the Firefox Accounts
deployment train cycle.
Sometimes,
[this makes me look like a dick](https://github.com/mozilla/fxa-activity-metrics/issues/57#issuecomment-283642668).

To work round this
(fixing the root cause
seems unrealistic),
I've created a dashboard
to be beamed onto every flat surface
in my immediate vicinity.
This project is the node module
that powers said dashboard.

## How do I install it?

Honestly,
it's not meant for you.

But if you insist:

```
npm i fxhey --save
```

## How do I use it?

### How do I load the library?

Use `require`:

```js
const fxhey = require('fxhey')
```

### How do I register for announcements?

Call `fxhey(callback, options)`,
where `callback` is a function
that will be invoked whenever
the deployed versions have changed:

```js
const cancel = fxhey(status => view.refresh(status))
```

Here,
`status` will be an object
that looks like this:

```js
{
  train: 81,           // The current FxA train number.
  time: 1488582720000, // Estimated deployment time, in milliseconds since the epoch, UTC.
  diffs: [             // Details of which servers have changed since the last announcement.
    { name: 'auth', current: { train: 81, patch: 2 }, previous: { train: 81, patch: 1 } }
  ],
  patches: [           // Details of which servers have been tagged with a patch level.
    { name: 'auth', train: 81, patch: 2 }
  ],
  versions: [          // Full version information for each deployed server.
    { name: 'content', train: 81, patch: 0, tag: 'v1.81.0', repo: 'mozilla/fxa-content-server' },
    { name: 'auth', train: 81, patch: 2, tag: 'v1.81.2', repo: 'mozilla/fxa-auth-server-private' },
    { name: 'profile', train: 79, patch: 0, tag: 'v0.79.0', repo: 'mozilla/fxa-profile-server' },
    { name: 'oauth', train: 81, patch: 0, tag: 'v1.81.0', repo: 'mozilla/fxa-oauth-server' }
  ]
}
```

After registering once,
your callback function will be called
whenever updated version data
is discovered
on the production servers.

### How do I cancel future announcements?

Just call the returned `cancel` function, like so:

```js
cancel()
```

### What options can I specify?

The options object looks like this:

```js
{
  rate: 1800000,       // Frequency, in milliseconds, that the servers will be checked. Defaults to 1 hour.
  immediate: false,    // Indicates whether an immediate callback is desired. Defaults to `true`.
  userAgent: 'Foo/1.0' // The user agent string used to identify requests to the server.
  status: {            // Initial status information used to determine whether the data has changed.
    train: 81,
    time: 1488582720000,
    versions: [
	  { train: 81, patch: 0 },
	  { train: 81, patch: 2 },
	  { train: 79, patch: 0 },
	  { train: 81, patch: 0 }
    ]
  }
}
```

## Is there a change log?

[Yes](CHANGELOG.md).

## How do I set up the dev environment?

To install the dependencies:

```
npm i
```

To run the unit tests:

```
npm t
```

To run a functional test against the production servers:

```
npm run tf
```

To lint the code:

```
npm run lint
```

## What license is it released under?

[MIT](LICENSE).



CHANGELOG.md
# Change log

## 0.2.3

### Bug fixes

* data: match new monorepo name (a14fcc7)

### Other changes

* package: update dependencies (49acd50)

## 0.2.2

### Other changes

* project: migrate to gitlab (2da1a20)

## 0.2.1

* fix: prevent undefined reference when no previous versions exist (b228acd)

## 0.2.0

* feature: add discrete time properties for each server (7bfdd77)
* chore: update dev dependencies (195dc11)

## 0.1.3

* fix: tolerate misordered initial status (b188c12)

## 0.1.2

* chore: rename from FxOi to FxHey (1a7e3dd)

## 0.1.1

* chore: update link text (ab7c406)
* fix: wow, travis is pedantic (abc166d)
* fix: add missing eslint dependency (66cefcf)
* fix: force immediate callback if set (ba9cec9)

## 0.1.0

* feature: fetch FxA version data from production servers (962b10f)
* feature: don't call back unless data has changed (edd7bfa)
* chore: add release script (c82ec18)
* feature: just return the repo name instead of the full url (e30e530)
* feature: add readme (44b196c)
* fix: eliminate lint errors (b5ab0e9)

## 0.0.0

feat: add project boilerplate



test.js
/* eslint-env mocha */

'use strict'

const { assert } = require('chai')
const pkg = require('./package.json')

suite('unit tests:', () => {
  const sinon = require('sinon')
  const proxyquire = require('proxyquire')
  const interval = 'foo'
  const version = {
    commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e',
    source: 'git://github.com/mozilla/fxa.git'
  }
  let timers, got, fxhey

  setup(() => {
    timers = {
      setInterval: sinon.spy(() => interval),
      clearInterval: sinon.spy()
    }
    // proxyquire barfs if got is a spied-on arrow function :-/
    // eslint-disable-next-line prefer-arrow-callback, brace-style
    got = sinon.spy(function () { return Promise.resolve({ body: version }) })
    fxhey = proxyquire('.', { timers, got })
  })

  test('interface is correct', () => {
    assert.isFunction(fxhey)
    assert.lengthOf(fxhey, 1)
  })

  test('timers were not invoked', () => {
    assert.equal(timers.setInterval.callCount, 0)
    assert.equal(timers.clearInterval.callCount, 0)
  })

  test('throws if rate is less than every half an hour', () => {
    assert.throws(() => fxhey(() => {}, { rate: 1799999 }))
  })

  test('does not throw if rate is every half an hour', () => {
    assert.doesNotThrow(() => fxhey(() => {}, { rate: 1800000 }))
  })

  test('throws if user agent is the empty string', () => {
    assert.throws(() => fxhey(() => {}, { userAgent: '' }))
  })

  test('does not throw if user agent is a non-empty string', () => {
    assert.doesNotThrow(() => fxhey(() => {}, { userAgent: 'wibble' }))
  })

  test('throws if initial train is zero', () => {
    assert.throws(() => fxhey(() => {}, {
      status: {
        train: 0,
        time: Date.now(),
        versions: [
          { name: 'content', train: 81, patch: 0 },
          { name: 'auth', train: 81, patch: 0 },
          { name: 'profile', train: 81, patch: 0 },
          { name: 'oauth', train: 81, patch: 0 }
        ]
      }
    }))
  })

  test('throws if initial time is zero', () => {
    assert.throws(() => fxhey(() => {}, {
      status: {
        train: 81,
        time: 0,
        versions: [
          { name: 'content', train: 81, patch: 0 },
          { name: 'auth', train: 81, patch: 0 },
          { name: 'profile', train: 81, patch: 0 },
          { name: 'oauth', train: 81, patch: 0 }
        ]
      }
    }))
  })

  test('throws if initial versions has length three', () => {
    assert.throws(() => fxhey(() => {}, {
      status: {
        train: 81,
        time: Date.now(),
        versions: [
          { name: 'content', train: 81, patch: 0 },
          { name: 'auth', train: 81, patch: 0 },
          { name: 'oauth', train: 81, patch: 0 }
        ]
      }
    }))
  })

  test('throws if version train is zero', () => {
    assert.throws(() => fxhey(() => {}, {
      status: {
        train: 81,
        time: Date.now(),
        versions: [
          { name: 'content', train: 81, patch: 0 },
          { name: 'auth', train: 81, patch: 0 },
          { name: 'profile', train: 81, patch: 0 },
          { name: 'oauth', train: 0, patch: 0 }
        ]
      }
    }))
  })

  test('throws if version train is greater than initial train', () => {
    assert.throws(() => fxhey(() => {}, {
      status: {
        train: 81,
        time: Date.now(),
        versions: [
          { name: 'content', train: 81, patch: 0 },
          { name: 'auth', train: 81, patch: 0 },
          { name: 'profile', train: 82, patch: 0 },
          { name: 'oauth', train: 81, patch: 0 }
        ]
      }
    }))
  })

  test('throws if version patch is negative', () => {
    assert.throws(() => fxhey(() => {}, {
      status: {
        train: 81,
        time: Date.now(),
        versions: [
          { name: 'content', train: 81, patch: 0 },
          { name: 'auth', train: 81, patch: -1 },
          { name: 'profile', train: 81, patch: 0 },
          { name: 'oauth', train: 81, patch: 0 }
        ]
      }
    }))
  })

  test('does not throw if initial status is valid', () => {
    assert.doesNotThrow(() => fxhey(() => {}, {
      status: {
        train: 81,
        time: Date.now(),
        versions: [
          { name: 'content', train: 81, patch: 0 },
          { name: 'auth', train: 81, patch: 0 },
          { name: 'profile', train: 81, patch: 0 },
          { name: 'oauth', train: 81, patch: 0 }
        ]
      }
    }))
  })

  test('does not throw without options', () => {
    assert.doesNotThrow(() => fxhey(() => {}))
  })

  test('throws if callback is not a function', () => {
    assert.throws(() => fxhey({}))
  })

  suite('fxhey with patch difference:', () => {
    let now, afterCallback, callback, cancel

    setup(done => {
      now = Date.now()
      sinon.stub(Date, 'now').callsFake(() => now)
      afterCallback = done
      callback = sinon.spy(() => afterCallback())
      version.version = '0.81.5'
      cancel = fxhey(callback, {
        status: {
          train: 81,
          time: now - 1,
          versions: [
            { name: 'content', train: 81, patch: 1, time: now - 1 },
            { name: 'auth', train: 81, patch: 2, time: now - 1 },
            { name: 'profile', train: 81, patch: 3, time: now - 1 },
            { name: 'oauth', train: 81, patch: 4, time: now - 1 }
          ]
        }
      })
    })

    teardown(() => {
      Date.now.restore()
    })

    test('returned function', () => {
      assert.isFunction(cancel)
      assert.lengthOf(cancel, 0)
    })

    test('called got correctly', () => {
      assert.equal(got.callCount, 4)
      const args = got.args[0]
      assert.lengthOf(args, 2)
      assert.equal(args[0], 'https://accounts.firefox.com/ver.json')
      assert.deepEqual(args[1], {
        json: true,
        headers: {
          'User-Agent': `FxHey/${pkg.version} (https://gitlab.com/philbooth/fxhey)`
        }
      })
      assert.equal(got.args[1][0], 'https://api.accounts.firefox.com/__version__')
      assert.equal(got.args[2][0], 'https://profile.accounts.firefox.com/__version__')
      assert.equal(got.args[3][0], 'https://oauth.accounts.firefox.com/__version__')
    })

    test('called callback correctly', () => {
      assert.equal(callback.callCount, 1)
      const args = callback.args[0]
      assert.lengthOf(args, 2)
      assert.isNull(args[0])
      assert.deepEqual(args[1], {
        time: now,
        train: 81,
        diffs: [
          { name: 'content', current: { train: 81, patch: 5 }, previous: { train: 81, patch: 1 } },
          { name: 'auth', current: { train: 81, patch: 5 }, previous: { train: 81, patch: 2 } },
          { name: 'profile', current: { train: 81, patch: 5 }, previous: { train: 81, patch: 3 } },
          { name: 'oauth', current: { train: 81, patch: 5 }, previous: { train: 81, patch: 4 } }
        ],
        patches: [
          { name: 'content', train: 81, patch: 5 },
          { name: 'auth', train: 81, patch: 5 },
          { name: 'profile', train: 81, patch: 5 },
          { name: 'oauth', train: 81, patch: 5 }
        ],
        versions: [
          { name: 'content', train: 81, patch: 5, tag: 'v0.81.5', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now },
          { name: 'auth', train: 81, patch: 5, tag: 'v0.81.5', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now },
          { name: 'profile', train: 81, patch: 5, tag: 'v0.81.5', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now },
          { name: 'oauth', train: 81, patch: 5, tag: 'v0.81.5', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now }
        ]
      })
    })

    test('called setInterval correctly', () => {
      assert.equal(timers.setInterval.callCount, 1)
      const args = timers.setInterval.args[0]
      assert.lengthOf(args, 2)
      assert.isFunction(args[0])
      assert.notEqual(args[0], callback)
      assert.equal(args[1], 3600000)
    })

    test('did not call clearInterval', () => {
      assert.equal(timers.clearInterval.callCount, 0)
    })

    suite('after interval without difference:', () => {
      setup(() => {
        afterCallback = () => {}
        timers.setInterval.args[0][0]()
      })

      test('called got four more times', () => {
        assert.equal(got.callCount, 8)
      })

      test('did not call callback', () => {
        assert.equal(callback.callCount, 1)
      })
    })

    suite('after interval with difference:', () => {
      setup(done => {
        afterCallback = done
        version.version = '0.81.6'
        timers.setInterval.args[0][0]()
      })

      test('called got four more times', () => {
        assert.equal(got.callCount, 8)
      })

      test('called callback correctly', () => {
        assert.equal(callback.callCount, 2)
        assert.notEqual(callback.args[1][1], callback.args[0][1])
        assert.deepEqual(callback.args[1][1], {
          time: now,
          train: 81,
          diffs: [
            { name: 'content', current: { train: 81, patch: 6 }, previous: { train: 81, patch: 5 } },
            { name: 'auth', current: { train: 81, patch: 6 }, previous: { train: 81, patch: 5 } },
            { name: 'profile', current: { train: 81, patch: 6 }, previous: { train: 81, patch: 5 } },
            { name: 'oauth', current: { train: 81, patch: 6 }, previous: { train: 81, patch: 5 } }
          ],
          patches: [
            { name: 'content', train: 81, patch: 6 },
            { name: 'auth', train: 81, patch: 6 },
            { name: 'profile', train: 81, patch: 6 },
            { name: 'oauth', train: 81, patch: 6 }
          ],
          versions: [
            { name: 'content', train: 81, patch: 6, tag: 'v0.81.6', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now },
            { name: 'auth', train: 81, patch: 6, tag: 'v0.81.6', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now },
            { name: 'profile', train: 81, patch: 6, tag: 'v0.81.6', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now },
            { name: 'oauth', train: 81, patch: 6, tag: 'v0.81.6', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now }
          ]
        })
      })
    })

    suite('cancel:', () => {
      setup(() => {
        cancel()
      })

      test('called clearInterval correctly', () => {
        assert.equal(timers.clearInterval.callCount, 1)
        const args = timers.clearInterval.args[0]
        assert.lengthOf(args, 1)
        assert.equal(args[0], interval)
      })
    })
  })

  suite('fxhey with version difference:', () => {
    let now, callback

    setup(done => {
      now = Date.now()
      sinon.stub(Date, 'now').callsFake(() => now)
      callback = sinon.spy(done)
      version.version = '1.82.0'
      fxhey(callback, {
        status: {
          train: 81,
          time: now - 1,
          versions: [
            { name: 'content', train: 81, patch: 0, time: now - 1 },
            { name: 'auth', train: 81, patch: 0, time: now - 1 },
            { name: 'profile', train: 81, patch: 0, time: now - 1 },
            { name: 'oauth', train: 81, patch: 0, time: now - 1 }
          ]
        }
      })
    })

    teardown(() => {
      Date.now.restore()
    })

    test('called callback correctly', () => {
      assert.equal(callback.callCount, 1)
      assert.deepEqual(callback.args[0][1], {
        time: now,
        train: 82,
        diffs: [
          { name: 'content', current: { train: 82, patch: 0 }, previous: { train: 81, patch: 0 } },
          { name: 'auth', current: { train: 82, patch: 0 }, previous: { train: 81, patch: 0 } },
          { name: 'profile', current: { train: 82, patch: 0 }, previous: { train: 81, patch: 0 } },
          { name: 'oauth', current: { train: 82, patch: 0 }, previous: { train: 81, patch: 0 } }
        ],
        patches: [],
        versions: [
          { name: 'content', train: 82, patch: 0, tag: 'v1.82.0', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now },
          { name: 'auth', train: 82, patch: 0, tag: 'v1.82.0', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now },
          { name: 'profile', train: 82, patch: 0, tag: 'v1.82.0', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now },
          { name: 'oauth', train: 82, patch: 0, tag: 'v1.82.0', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now }
        ]
      })
    })
  })

  suite('fxhey with partial difference:', () => {
    let now, callback

    setup(done => {
      now = Date.now()
      sinon.stub(Date, 'now').callsFake(() => now)
      callback = sinon.spy(done)
      version.version = '1.82.0'
      fxhey(callback, {
        status: {
          train: 82,
          time: now - 1,
          versions: [
            { name: 'content', train: 82, patch: 0, time: now - 1 },
            { name: 'auth', train: 82, patch: 0, time: now - 1 },
            { name: 'profile', train: 82, patch: 0, time: now - 1 },
            { name: 'oauth', train: 81, patch: 0, time: now - 1 }
          ]
        }
      })
    })

    teardown(() => {
      Date.now.restore()
    })

    test('called callback correctly', () => {
      assert.equal(callback.callCount, 1)
      assert.deepEqual(callback.args[0][1], {
        time: now,
        train: 82,
        diffs: [
          { name: 'oauth', current: { train: 82, patch: 0 }, previous: { train: 81, patch: 0 } }
        ],
        patches: [],
        versions: [
          { name: 'content', train: 82, patch: 0, tag: 'v1.82.0', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now - 1 },
          { name: 'auth', train: 82, patch: 0, tag: 'v1.82.0', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now - 1 },
          { name: 'profile', train: 82, patch: 0, tag: 'v1.82.0', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now - 1 },
          { name: 'oauth', train: 82, patch: 0, tag: 'v1.82.0', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now }
        ]
      })
    })
  })

  suite('fxhey with no differences:', () => {
    let now, callback

    setup(() => {
      now = Date.now()
      sinon.stub(Date, 'now').callsFake(() => now)
      callback = sinon.spy()
      version.version = '1.81.0'
      fxhey(callback, {
        status: {
          train: 81,
          time: now - 1,
          versions: [
            { name: 'content', train: 81, patch: 0 },
            { name: 'auth', train: 81, patch: 0 },
            { name: 'profile', train: 81, patch: 0 },
            { name: 'oauth', train: 81, patch: 0 }
          ]
        }
      })
    })

    teardown(() => {
      Date.now.restore()
    })

    test('called got four times', () => {
      assert.equal(got.callCount, 4)
    })

    test('did not call callback', () => {
      assert.equal(callback.callCount, 0)
    })
  })

  suite('fxhey with version and ordering differences:', () => {
    let now, callback

    setup(done => {
      now = Date.now()
      sinon.stub(Date, 'now').callsFake(() => now)
      callback = sinon.spy(done)
      version.version = '1.82.0'
      fxhey(callback, {
        status: {
          train: 81,
          time: now - 1,
          versions: [
            { name: 'oauth', train: 81, patch: 1, time: now - 1 },
            { name: 'profile', train: 81, patch: 2, time: now - 1 },
            { name: 'auth', train: 81, patch: 3, time: now - 1 },
            { name: 'content', train: 81, patch: 4, time: now - 1 }
          ]
        }
      })
    })

    teardown(() => {
      Date.now.restore()
    })

    test('called callback correctly', () => {
      assert.equal(callback.callCount, 1)
      assert.deepEqual(callback.args[0][1], {
        time: now,
        train: 82,
        diffs: [
          { name: 'content', current: { train: 82, patch: 0 }, previous: { train: 81, patch: 4 } },
          { name: 'auth', current: { train: 82, patch: 0 }, previous: { train: 81, patch: 3 } },
          { name: 'profile', current: { train: 82, patch: 0 }, previous: { train: 81, patch: 2 } },
          { name: 'oauth', current: { train: 82, patch: 0 }, previous: { train: 81, patch: 1 } }
        ],
        patches: [],
        versions: [
          { name: 'content', train: 82, patch: 0, tag: 'v1.82.0', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now },
          { name: 'auth', train: 82, patch: 0, tag: 'v1.82.0', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now },
          { name: 'profile', train: 82, patch: 0, tag: 'v1.82.0', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now },
          { name: 'oauth', train: 82, patch: 0, tag: 'v1.82.0', commit: '75ca755f94be44c06c55fab8e3fccfedb0e4b59e', repo: 'mozilla/fxa', time: now }
        ]
      })
    })
  })
})

suite('functional test:', () => {
  let now, cancel, error, result

  setup(done => {
    now = Date.now()
    cancel = require('.')((e, r) => {
      error = e
      result = r
      done()
    }, {
      status: {
        time: Date.UTC(2017, 0, 1),
        train: 81,
        versions: [
          { name: 'content', train: 81, patch: 0 },
          { name: 'auth', train: 81, patch: 0 },
          { name: 'profile', train: 81, patch: 0 },
          { name: 'oauth', train: 81, patch: 0 }
        ]
      }
    })
  })

  teardown(() => {
    cancel()
  })

  test('result seems correct', () => {
    assert.isNull(error)
    assert.isObject(result)
    assert.isNumber(result.train)
    assert.isAtLeast(result.train, 81)
    assert.isNumber(result.time)
    assert.isAtLeast(result.time, now)
    assert.isArray(result.diffs)
    assert.isArray(result.patches)
    assert.isArray(result.versions)
    assert.lengthOf(result.versions, 4)
    let patchCount = 0
    result.versions.forEach(version => {
      assert.isNumber(version.train)
      assert.isAtMost(version.train, result.train)
      assert.isNumber(version.patch)
      assert.isAtLeast(version.patch, 0)
      if (version.patch > 0) {
        ++patchCount
      }
      assert.match(version.tag, new RegExp(`^v[01]\\.${version.train}\\.${version.patch}$`))
      assert.match(version.commit, /^[a-f0-9]{40}$/)
    })
    assert.lengthOf(result.patches, patchCount)
    assert.equal(result.versions[0].name, 'content')
    assert.equal(result.versions[0].repo, 'mozilla/fxa-content-server')
    assert.equal(result.versions[1].name, 'auth')
    assert.equal(result.versions[1].repo, 'mozilla/fxa-auth-server-private')
    assert.equal(result.versions[2].name, 'profile')
    assert.equal(result.versions[2].repo, 'mozilla/fxa-profile-server')
    assert.equal(result.versions[3].name, 'oauth')
    assert.equal(result.versions[3].repo, 'mozilla/fxa-oauth-server')
  })
})



index.js
'use strict'

const { assert } = require('check-types')
const got = require('got')
const { setInterval, clearInterval } = require('timers')
const pkg = require('./package.json')

const DEFAULT_RATE = 1000 * 60 * 60
const MINIMUM_RATE = DEFAULT_RATE / 2
const DEFAULT_USER_AGENT = `FxHey/${pkg.version} (https://gitlab.com/philbooth/fxhey)`
const REPO_MATCH = /mozilla\/fxa(?:-private)?/
const SERVERS = [
  { name: 'content', url: 'https://accounts.firefox.com/ver.json' },
  { name: 'auth', url: 'https://api.accounts.firefox.com/__version__' },
  { name: 'profile', url: 'https://profile.accounts.firefox.com/__version__' },
  { name: 'oauth', url: 'https://oauth.accounts.firefox.com/__version__' }
]

let previousStatus

module.exports = (
  callback,
  {
    rate = DEFAULT_RATE,
    immediate = true,
    userAgent = DEFAULT_USER_AGENT,
    status = {
      train: 81,
      time: Date.UTC(2017, 2, 3, 23, 12),
      versions: [
        { train: 81, patch: 0 },
        { train: 81, patch: 2 },
        { train: 79, patch: 0 },
        { train: 81, patch: 0 }
      ]
    }
  } = {}
) => {
  assert.function(callback)
  assert.greaterOrEqual(rate, MINIMUM_RATE)
  assert.nonEmptyString(userAgent)
  assert.object(status)
  assert.greater(status.train, 0)
  assert.positive(status.time)
  assert.array.of.object(status.versions)
  assert.hasLength(status.versions, 4)
  status.versions.forEach(version => {
    assert.positive(version.train)
    assert.lessOrEqual(version.train, status.train)
    assert.greaterOrEqual(version.patch, 0)
  })

  previousStatus = clone(status)

  if (immediate) {
    getStatus(userAgent, callback, true)
  }

  let interval = setInterval(getStatus.bind(null, userAgent, callback), rate)

  return () => {
    if (interval) {
      clearInterval(interval)
      interval = null
    }
  }
}

function getStatus (userAgent, callback, forceCallback) {
  return fetchVersions(userAgent)
    .then(generateStatus)
    .then(
      status => {
        previousStatus = clone(status)
        if (status.diffs.length > 0 || forceCallback) {
          callback(null, status)
        }
      },
      error => callback(error, clone(previousStatus))
    )
}

function fetchVersions (userAgent) {
  return Promise.all(SERVERS.map(server => {
    return got(server.url, { json: true, headers: { 'User-Agent': userAgent } })
      .then(result => {
        const body = result.body
        return {
          name: server.name,
          version: body.version,
          repo: parseRepo(body.source),
          tag: `v${body.version}`,
          commit: body.commit
        }
      })
  }))
}

function parseRepo (source) {
  return REPO_MATCH.exec(source)[0]
}

function generateStatus (versions) {
  const result = versions.reduce((status, version) => {
    const { train, patch } = parseVersion(version.version)
    const { name, tag, commit, repo } = version
    let previousVersion
    let time
    previousStatus.versions.some(candidateVersion => {
      if (candidateVersion.name === name) {
        previousVersion = candidateVersion
        return true
      }
      return false
    })
    const { train: previousTrain, patch: previousPatch } = previousVersion || {}

    if (train > status.train) {
      status.train = train
    }

    if (train === previousTrain && patch === previousPatch) {
      time = previousVersion.time
    } else {
      if (status.time === previousStatus.time) {
        status.time = Date.now()
      }
      time = status.time
      status.diffs.push({
        name,
        current: { train, patch },
        previous: { train: previousTrain, patch: previousPatch }
      })
    }

    if (patch > 0 && train === status.train) {
      status.patches.push({ name, train, patch })
    }

    status.versions.push({ name, train, patch, tag, commit, repo, time })

    return status
  }, {
    train: previousStatus.train,
    time: previousStatus.time,
    diffs: [],
    patches: [],
    versions: []
  })

  result.patches = result.patches.filter(patch => patch.train === result.train)

  return result
}

function parseVersion (version) {
  const split = version.split('.')
  return {
    train: parseInt(split[1]),
    patch: parseInt(split[2])
  }
}

function clone (value) {
  if (! value || typeof value !== 'object') {
    return value
  }

  if (Array.isArray(value)) {
    return value.map(item => clone(item))
  }

  return Object.keys(value).reduce((cloned, key) => {
    cloned[key] = clone(value[key])
    return cloned
  }, {})
}

